### PR TITLE
Merge W3C caps into JSONWP desiredCapabilities

### PR DESCRIPTION
**Pull request by dpgraham**

### CODE REVIEW

File: lib/appium.js [hunk 1]
```diff
@@ -172,9 +172,18 @@ class AppiumDriver extends BaseDriver {
...
-    if (capabilities) {
-      desiredCaps = processCapabilities(capabilities, null, false);
+  /**
+   * Create a new session
+   * @param {Object} desiredCaps JSONWP formatted desired capabilities
+   * @param {Object} reqCaps Required capabilities
+   * @param {Object} capabilities W3C capabilities
+   */
+  async createSession (desiredCaps={}, reqCaps, capabilities) {
+    if (_.isObject(capabilities)) {
```
Comment (mykola-mokhnach) [inline lib/appium.js]: -> isPlainObject

File: lib/appium.js [hunk 2]
```diff
@@ -172,9 +172,18 @@ class AppiumDriver extends BaseDriver {
...
   }
 
-  async createSession (desiredCaps, reqCaps, capabilities) {
-    if (capabilities) {
-      desiredCaps = processCapabilities(capabilities, null, false);
+  /**
+   * Create a new session
+   * @param {Object} desiredCaps JSONWP formatted desired capabilities
+   * @param {Object} reqCaps Required capabilities
+   * @param {Object} capabilities W3C capabilities
```
Comment (mykola-mokhnach) [inline lib/appium.js]: Please add a description of the returned result

File: lib/appium.js [hunk 3]
```diff
@@ -172,9 +172,18 @@ class AppiumDriver extends BaseDriver {
...
+   */
+  async createSession (desiredCaps={}, reqCaps, capabilities) {
+    if (_.isObject(capabilities)) {
+      // Merging W3C caps into desiredCaps is a stop-gap until all the clients and drivers become fully W3C compliant
+      log.info(`Merged W3C capabilities ${_.truncate(JSON.stringify(capabilities), 20)} into desiredCapabiilities object ${JSON.stringify(desiredCaps)}`);
```
Comment (mykola-mokhnach) [inline lib/appium.js]: is 20 chars enough?

Comment (imurchie) [reply lib/appium.js]: This will actually truncate to 30, the default, because the usage is `_.truncate(JSON.stringify(capabilities), {length: 20})`

Comment (imurchie) [reply lib/appium.js]: Type: `desiredCapabiilities`

File: lib/appium.js (hunk 1 above)
Comment (imurchie) [reply lib/appium.js]: What is the purpose of the check? What would the value be other than an object?

**Review by imurchie — CHANGES_REQUESTED** [review]
Approach seems good. A couple of issues.

**Comment by dpgraham** [comment]
Fixes pushed.

File: lib/appium.js [hunk 4]
```diff
@@ -172,9 +172,19 @@ class AppiumDriver extends BaseDriver {
...
+   * @param {Object} desiredCaps JSONWP formatted desired capabilities
+   * @param {Object} reqCaps Required capabilities
+   * @param {Object} capabilities W3C capabilities
+   * @return {Array} Unique session ID and capabilities
+   */
+  async createSession (desiredCaps={}, reqCaps, capabilities) {
     if (capabilities) {
-      desiredCaps = processCapabilities(capabilities, null, false);
+      // Merging W3C caps into desiredCaps is a stop-gap until all the clients and drivers become fully W3C compliant
+      log.info(`Merged W3C capabilities ${_.truncate(JSON.stringify(capabilities), {length: 50})} into desiredCapabiilities object ${_.truncate(JSON.stringify(desiredCaps), {length: 50})}`);
```
Comment (imurchie) [inline lib/appium.js]: Typo: `desiredCapabiilities`.

**Review by imurchie — APPROVED** [review]

**Pull request merged**